<a href="https://colab.research.google.com/github/AlbericoJose/Coleta-e-Analise/blob/main/Coleta_dados_vacina%C3%A7%C3%A3o.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Coleta de dados estruturados

Os dados estruturados são provinientes de estruturas definidas, inflexivel, pensadas antes mesmo da própria criação dos dados. Exemplo de dados estruturados: Tabelas de banco de dados, planilhas eletronicas, arquivos de textos como txt e csv.

A base de dados está disponível no seguinte link: https://www.kaggle.com/datasets/gpreda/covid-world-vaccination-progress


o dataset utilizado nesse estudo foi coletado no dia 26/02/2021 às 17:50

In [ ]:
#importar as bibliotecas
import pandas as pd
import os

# Importando os dados

O primeiro passo é encontrar o local aonde o arquivo a ser trabalhao vai estar amazenado. Após isso, vamos abri-lo e em seguida vamos adiciona-lo dentro de um dataframe do pandas

In [ ]:
nome_arquivo = 'country_vaccinations.csv'
#Usando no jupytet notebook
#Dir = "C:\\Projetos\\coleta_dados\\dataset"
#CaminhoDados = os.path.join(Dir,nome_arquivo)

# lendo dados
dados = pd.read_csv(nome_arquivo, sep =',' , encoding='latin-1')


In [ ]:
# validamos o caminho
#CaminhoDados

In [ ]:
dados # apresenta os dados

Criando o dataframe no pandas e ler o arquivo na pasta de origem. Para abrir o arquivo vamos utilizar o comando read_csv e utilizaro separador ',' de acordo com o dado fornecido. Além disso, temos que utilizar o encoding informando a codificação do arquivo.

In [ ]:
#Usando no jupytet notebook
#dataframe = pd.read_csv(dados, sep =',' , encoding='latin-1')
dataframe = dados;

In [ ]:
# exibe os 5 primeiros registros
dataframe.head(5)

country iso_code        date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG  2021-02-22                 0.0                0.0   
1  Afghanistan      AFG  2021-02-23                 NaN                NaN   
2  Afghanistan      AFG  2021-02-24                 NaN                NaN   
3  Afghanistan      AFG  2021-02-25                 NaN                NaN   
4  Afghanistan      AFG  2021-02-26                 NaN                NaN   

   people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                      NaN                     NaN                 NaN   
1                      NaN                     NaN              1367.0   
2                      NaN                     NaN              1367.0   
3                      NaN                     NaN              1367.0   
4                      NaN                     NaN              1367.0   

   total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                             0.0                            0.0   
1                             NaN                            NaN   
2                             NaN                            NaN   
3                             NaN                            NaN   
4                             NaN                            NaN   

   people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
0                                  NaN                             NaN   
1                                  NaN                            34.0   
2                                  NaN                            34.0   
3                                  NaN                            34.0   
4                                  NaN                            34.0   

                                            vaccines  \
0  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
1  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
2  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
3  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
4  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   

                 source_name            source_website  
0  World Health Organization  https://covid19.who.int/  
1  World Health Organization  https://covid19.who.int/  
2  World Health Organization  https://covid19.who.int/  
3  World Health Organization  https://covid19.who.int/  
4  World Health Organization  https://covid19.who.int/

#Analisando os dados

O dataset utilizado é referente a estatistica do número de vacinados da COVID-19 em todos os paises do mundo.

Para esse estudo vamos utilizar as variaveis: coutry = pais; date = data; total_vaccinations = nr do total de vacinados; daily_vaccinations = nr de vacinados por dia; vaccines = nome das vacinas utilizada no pais;

Vamos responder as perguntas:


1.   Qual os pais com maior nr de pessoas imunizadas?
2.   Qual vacina é a mais utilizada no mundo? ** Existem diferentes tipos de vacinas disponivel no mundo.
3. Qual pais possui o maior nr de vacinados por dia?
4. Qual a média de vacinação diariamente por país? 


Após responder as perguntas vamos exportat as respostas obtidas.

In [ ]:
#verifica dados do Brasil p/ dia
dataframe.query('country == "Brazil"').sort_values('date', ascending=False)[:2]

country iso_code        date  total_vaccinations  people_vaccinated  \
11085  Brazil      BRA  2022-03-29         413559595.0        181078067.0   
11084  Brazil      BRA  2022-03-28         412177025.0        180881628.0   

       people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
11085              160272858.0               1382570.0            656803.0   
11084              160008924.0                628823.0            546999.0   

       total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
11085                          193.26                          84.62   
11084                          192.61                          84.53   

       people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
11085                                74.90                          3069.0   
11084                                74.77                          2556.0   

                                                vaccines  \
11085  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11084  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   

                                           source_name  \
11085  State governments via coronavirusbra1.github.io   
11084  State governments via coronavirusbra1.github.io   

                          source_website  
11085  https://coronavirusbra1.github.io  
11084  https://coronavirusbra1.github.io

In [ ]:
# verifica o nr de paises distintos na base
dataframe['country'].nunique()

223

**Observem que temos 223 registros de paises no nosso datatset**

#1.   Qual os pais com maior nr de pessoas imunizadas?

Para que a análise seja realizada de forma correta é necessário analisar profundamente os dados. Nessa etapa, precisamos fazer vários testes afim de conhecer melhor os dados. Perceba que o dataset traz informações por data de coleta. Dessa forma, temos que filtrar p ultimo registr baseado na data.

Será que essa análise é correta?


In [ ]:
# identificando a data mais recente da coleta
ultima_data = dataframe['date'].max()
ultima_data

'2022-03-29'

In [ ]:
# filtrando a ultima data
ultima_data = dataframe[dataframe.date == ultima_data]

In [ ]:
#O comando shape mostra a quantidade de linhas e colunas detro de um dataset
ultima_data.shape

(54, 15)

#Atenção!

Perceba que o dataset filtrado pela ultima data retornou (54,15) ou seja 54 linhas e 15 colunas. Isso não esta certo! Pois verificamos que temos X oaises no dataset original.

Dessa forma, ao filtrar pela última data estamos excluindo registros de outros paises que não foram ainda computados na última coleta.

para resolver esse problema, temos que selecionar o último registro de data de cada pais. Assim não vamos deixar nenhum país de fora.

Vamos fazer um agrupamento do pais e selecionar a ultima data de atualização que foi realizada.

In [ ]:
dataframe.groupby('country')['date'].max().sort_values(ascending = False)

country
Lebanon                            2022-03-29
Lithuania                          2022-03-29
Kosovo                             2022-03-29
Kazakhstan                         2022-03-29
Japan                              2022-03-29
                                      ...    
Pitcairn                           2021-09-07
Turkmenistan                       2021-09-05
Bonaire Sint Eustatius and Saba    2021-09-01
Saint Helena                       2021-05-05
Falkland Islands                   2021-04-14
Name: date, Length: 223, dtype: object

Vimos que existem paises que não possuem a atualização da ultima data da coelta. Dessa forma, temos que procurar um aoutra maneira para obter esses dados. A variável total_vaccinations contém o total de vacinações naquele pais, ou seja, podemos utilizar como parametro para buscar a ultima data de coleta. Desta forma, vamos agrupar por país e pegar o valor máximo obtido nessa variável.

In [ ]:
dataframe.groupby('country')['total_vaccinations'].max().sort_values(ascending = False)

country
China               3.263129e+09
India               1.834501e+09
United States       5.601818e+08
Brazil              4.135596e+08
Indonesia           3.771089e+08
                        ...     
Falkland Islands    4.407000e+03
Montserrat          4.211000e+03
Niue                4.161000e+03
Tokelau             1.936000e+03
Pitcairn            9.400000e+01
Name: total_vaccinations, Length: 223, dtype: float64

Vamos então listar os Top 10 paises

In [ ]:
nr_imunizados_pais =  dataframe.groupby('country')['total_vaccinations'].max().sort_values(ascending = False)
nr_imunizados_pais

country
China               3.263129e+09
India               1.834501e+09
United States       5.601818e+08
Brazil              4.135596e+08
Indonesia           3.771089e+08
                        ...     
Falkland Islands    4.407000e+03
Montserrat          4.211000e+03
Niue                4.161000e+03
Tokelau             1.936000e+03
Pitcairn            9.400000e+01
Name: total_vaccinations, Length: 223, dtype: float64

In [ ]:
type(nr_imunizados_pais)

pandas.core.series.Series

Percebam que efetuar uma operação, o seu resultado se transforma em uma série pandas.
Uma série pandas se caracterizada por um array unidimencional capaz de armazenar qualquer tipo de dados com rótulos ou índice de eixo. Um exmplo de  um objeto séries e uma coluna de um Dataframe.

Para transformar novamente em um dataframe temos que usar o comando to_frame() e resetar o index.

In [ ]:
nr_imunizados_pais =  nr_imunizados_pais.to_frame().reset_index()
nr_imunizados_pais


country  total_vaccinations
0               China        3.263129e+09
1               India        1.834501e+09
2       United States        5.601818e+08
3              Brazil        4.135596e+08
4           Indonesia        3.771089e+08
..                ...                 ...
218  Falkland Islands        4.407000e+03
219        Montserrat        4.211000e+03
220              Niue        4.161000e+03
221           Tokelau        1.936000e+03
222          Pitcairn        9.400000e+01

[223 rows x 2 columns]

Agora temos a realação do nr total de imunizados por país.

Vamos utilizar o metodo sort_values na variável total_vaccinations para ordenar e o parâmetri ascending como False para ordenar, do maior para o menor. Alem de selecionar os 10 primeiros países

In [ ]:
nr_imunizados_pais.sort_values('total_vaccinations', ascending=False).head(10)

country  total_vaccinations
0          China        3.263129e+09
1          India        1.834501e+09
2  United States        5.601818e+08
3         Brazil        4.135596e+08
4      Indonesia        3.771089e+08
5          Japan        2.543456e+08
6     Bangladesh        2.436427e+08
7       Pakistan        2.193686e+08
8        Vietnam        2.031444e+08
9         Mexico        1.919079e+08

#2.Qual vacina é a mais utilizada no mundo? 
Para descobrimos, devemos criar um dataframe contendo as vacinas utilizads por país, com isso teremos uma relação com todos os paises e suas vacinas em uma determinada data.

In [ ]:
nr_vacina_pais = dataframe[['country','date','vaccines']]
nr_vacina_pais

country        date  \
0      Afghanistan  2021-02-22   
1      Afghanistan  2021-02-23   
2      Afghanistan  2021-02-24   
3      Afghanistan  2021-02-25   
4      Afghanistan  2021-02-26   
...            ...         ...   
86507     Zimbabwe  2022-03-25   
86508     Zimbabwe  2022-03-26   
86509     Zimbabwe  2022-03-27   
86510     Zimbabwe  2022-03-28   
86511     Zimbabwe  2022-03-29   

                                                vaccines  
0      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
1      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
2      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
3      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
4      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
...                                                  ...  
86507  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86508  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86509  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86510  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86511  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  

[86512 rows x 3 columns]

Iremos agrupar as vacinas por pais. Se no dataset tiver um pais utilizando duas vacinas, ele irá contabilizar.

In [ ]:
nr_vacina_pais.groupby('vaccines')['country'].nunique().sort_values(ascending=False)

vaccines
Oxford/AstraZeneca                                                                                    20
Johnson&Johnson, Moderna, Oxford/AstraZeneca, Pfizer/BioNTech                                         17
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech                                                          15
Oxford/AstraZeneca, Pfizer/BioNTech                                                                   11
Johnson&Johnson, Moderna, Novavax, Oxford/AstraZeneca, Pfizer/BioNTech                                 8
                                                                                                      ..
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinopharm/Beijing, Sinovac                               1
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinovac, Sputnik Light, Sputnik V, ZF2001                1
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinovac, Sputnik V                                       1
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, 

In [ ]:
vacinas_pais =  dataframe.groupby('country')['vaccines'].nunique()
vacinas_pais = vacinas_pais.to_frame().reset_index()

Temos a relação entre as vacinas mais utilizadas no mundo.  Notem que a vacina Oxford/AstraZeneca contém em 20 paises.

#3. Qual pais possui o maior nr de vacinados por dia?

Para obter a resposta temos que analizar a variável daily_vaccinations.

In [ ]:
dataframe.groupby('country')['daily_vaccinations'].max().sort_values(ascending=False)

country
China            22424286.0
India            10037995.0
Bangladesh        3758404.0
United States     3506960.0
Pakistan          2175773.0
                    ...    
Saint Helena          165.0
Niue                   87.0
Montserrat             53.0
Tokelau                23.0
Pitcairn                1.0
Name: daily_vaccinations, Length: 223, dtype: float64

#4. Qual a média de vacinação diariamente por país? 

In [ ]:
media_pais = dataframe.groupby('country')['daily_vaccinations'].mean().sort_values(ascending=False)
media_pais = media_pais.to_frame().reset_index()
media_pais

country  daily_vaccinations
0                China        6.930368e+06
1                India        4.175994e+06
2        United States        1.191727e+06
3               Brazil        9.435287e+05
4            Indonesia        8.462893e+05
..                 ...                 ...
218  Wallis and Futuna        3.388649e+01
219            Tokelau        1.739823e+01
220               Niue        1.631698e+01
221         Montserrat        1.053659e+01
222           Pitcairn        5.119048e-01

[223 rows x 2 columns]

Todas as perguntas respondidas!!!
Um dos trabalho de um cientista de dados é extrair, limpar, tratr e gerar novos dados.

Geramos alguns dataframes das perguntas que criamos.
1. nr_imunizados_pais
2. nr_vacina_pais
3. media_pais

Agora vamos criar uma noba base de dados utilizando os dados que tratamos anteriomente.


In [ ]:
# renomenado as colunas
media_pais = media_pais.rename(columns={'country':'pais','daily_vaccination':'media_vacinacao'})
media_pais[:5]

In [ ]:
# renomenado as colunas
nr_vacina_pais = nr_vacina_pais.rename(columns={'country':'pais','vaccines':'vacina'})
nr_vacina_pais[:5]

In [ ]:
nr_imunizados_pais = nr_imunizados_pais.rename(columns={'country':'pais','total_vaccinations':'nr_imunizados'})
nr_imunizados_pais[:5] 


pais  nr_imunizados
0          China   3.263129e+09
1          India   1.834501e+09
2  United States   5.601818e+08
3         Brazil   4.135596e+08
4      Indonesia   3.771089e+08

# Criando um novo dataset
para criar um novo dateser, vamos utilizar a função merge. Essa função vai comparar os campos das tabelas que são chaves iguais, como o nome do pais e realizar uma junção entre as tabelas. Assim vamos ter todos os dados tratados em uma única tabela.

In [ ]:
db_tratados = pd.merge(nr_imunizados_pais,media_pais, how ='inner')

In [ ]:
db_tratados = pd.merge(db_tratados,nr_imunizados_pais)

Podemos também realizar de forma direta conforme o script abaixo.

In [ ]:
pd.merge(pd.merge(nr_imunizados_pais,media_pais),nr_vacina_pais)

pais  nr_imunizados  daily_vaccinations        date  \
0         China   3.263129e+09        6.930368e+06  2020-12-15   
1         China   3.263129e+09        6.930368e+06  2020-12-16   
2         China   3.263129e+09        6.930368e+06  2020-12-17   
3         China   3.263129e+09        6.930368e+06  2020-12-18   
4         China   3.263129e+09        6.930368e+06  2020-12-19   
...         ...            ...                 ...         ...   
86507  Pitcairn   9.400000e+01        5.119048e-01  2021-09-03   
86508  Pitcairn   9.400000e+01        5.119048e-01  2021-09-04   
86509  Pitcairn   9.400000e+01        5.119048e-01  2021-09-05   
86510  Pitcairn   9.400000e+01        5.119048e-01  2021-09-06   
86511  Pitcairn   9.400000e+01        5.119048e-01  2021-09-07   

                                                  vacina  
0      CanSino, Sinopharm/Beijing, Sinopharm/Wuhan, S...  
1      CanSino, Sinopharm/Beijing, Sinopharm/Wuhan, S...  
2      CanSino, Sinopharm/Beijing, Sinopharm/Wuhan, S...  
3      CanSino, Sinopharm/Beijing, Sinopharm/Wuhan, S...  
4      CanSino, Sinopharm/Beijing, Sinopharm/Wuhan, S...  
...                                                  ...  
86507                                 Oxford/AstraZeneca  
86508                                 Oxford/AstraZeneca  
86509                                 Oxford/AstraZeneca  
86510                                 Oxford/AstraZeneca  
86511                                 Oxford/AstraZeneca  

[86512 rows x 5 columns]

#Exportando dados tratados

Exportando para um arquivo txt.

In [ ]:
#Exportando para um arquivo txt.
db_tratados.to_csv('estatictica_pais.csv', index = True, sep =';')

In [ ]:
#Exportando para um arquivo json.
db_tratados.to_json('estatictica.json', orient='records')